_posts/2020-11-30-flexible-and-safe-tuples-in-scala3.md

```diff
@@ -0,0 +1,387 @@
+```scala
+(Set(0), 1, "2", 3.0, List(4)).splitAt(3) // ((Set(0), 1, "2", 3.0), (3.0, List(4)))
+```
+
+### Tuples can transform
+
+Again, similarly to conversion methods on collections, it is possible to
+transform a tuple into another collection.
+
+We have to pay attention to the type of the resulting collection.
+Let's start with the simple case: as its name might hint,
+`toArray` produces an array. The type of its elements will always be
+`Object`. This makes it easy to reason about this method although it
+forgets the type of the elements.
+It is also possible to use `.toIArray` which has exactly the same behavior
+but produces an `IArray` where the `I` stands for immutable.
+```scala
+scala> (1, "2").toArray
+val res0: Array[Object] = Array(1, 2)
+```
+
+I believe however that the most interesting conversion is `toList`
+which produces yes a list, but not a `List[Any]`.
+Instead it creates `List[U]` where `U` is the union type of the types
+of the elements of the tuple.
+That is:
+
+```scala
+(1, "2", 3.0)
+val res0: List[Int | String | Double] = (1, "2", 3.0).toList
+```
+This is interesting because the type information is somehow maintained.
+We can iterate over `list` and use pattern matching to apply the
+correct transformation, knowing exactly how many and what cases to
+treat:
+
+```scala
+// The compiler tells he cannot help with checking:
+// Non-exhaustive match
+(1, "2").toArray.map {
+  case i: Int => (i * 2).toString
+  case j: String => j
+}
+
+// The code compiles without errors or warning
+// the compile verified that we handled all possible cases
+(1, "2").toList.map {
+  case i: Int => (i + 2).toString
+  case j: String => j
+}
+```
+Read more about [Union types here](https://dotty.epfl.ch/docs/reference/new-types/union-types.html).
+
+We can also transform a tuple by appling a function to each element.
+The method, similarly to what we are used to with collections, is called
+`map`. The difference from collections (or functors) is however
+that they expect a `f: A => B` where `A` is the type of the elements
+of the collection.
+With tuples each element has a different type!
+How can we generalize the concept of a function whose argument type is
+not fixed ?
+We can use a **`PolyFunction`**. This is a more advanced syntax:
+
+```scala
+(1, 'a', "dog", 3.0).map[[X] =>> Option[String]]([T] => (t: T) => t match {
+  case i: String => i.toUpperCase
+  case j: Int => j.toString
+})
+```
+
+## Tuples can compose
+
+The last operator allows to pair the elements of two tuples.
+You might have guessed, it is called `zip`. If the two tuples have
+different lengths, the extra elements of the longest will be
+ignored:
+
+```scala
+val numbers = (1, 2, 3, 4, 5)
+val letters = ('a', 'b', 'c')
+
+numbers.zip(letters) // ((1, 'a'), (2, 'b'), (3, 'c'))
+```
+
+# Under the hood: new type operators of Scala 3
+
+I believe that the core new features that allows such a flexible
+implementation of tuples are **match types**.
```

Comment:
Earlier in the blog post it said "compiler feature", and I wondered what that meant.

Now, some pages later, I'm finding out that it refers mainly to match types. I'd suggest simply saying so in the intro, rather than keep the reader in suspense.

Also I would call match types a language feature, not a "compiler feature". (`-Xlint` and `-rewrite` are examples of compiler features.)

Reply:
I totally agree, compiler feature is not the appropriate expression. Thanks for catching it.
Some suspense is nice in all stories no ? :smile:

Reply:
I leave this same feedback about "suspense" on basically everyone's blog posts :-)